src/main/portfolio_driver.cpp

```diff
@@ -0,0 +1,516 @@
+#include "base/output.h"
+#include "main/command_executor.h"
+#include "parser/parser.h"
+#include "smt/command.h"
+
+namespace cvc5::main {
+
+bool ExecutionContext::solveContinuous(parser::Parser* parser,
+                                       bool stopAtSetLogic)
+{
+  std::unique_ptr<cvc5::Command> cmd;
+  bool interrupted = false;
+  bool status = true;
+  while (status)
+  {
+    if (interrupted)
+    {
+      solver().getDriverOptions().out() << cvc5::CommandInterrupted();
+      d_executor->reset();
+      break;
+    }
+    cmd.reset(parser->nextCommand());
+    if (cmd == nullptr) break;
+
+    status = d_executor->doCommand(cmd);
+    if (cmd->interrupted() && status == 0)
+    {
+      interrupted = true;
+      break;
+    }
+
+    if (dynamic_cast<cvc5::QuitCommand*>(cmd.get()) != nullptr)
+    {
+      break;
+    }
+    if (stopAtSetLogic)
+    {
+      auto* slc = dynamic_cast<cvc5::SetBenchmarkLogicCommand*>(cmd.get());
+      if (slc != nullptr)
+      {
+        d_logic = slc->getLogic();
+        break;
+      }
+    }
+  }
+  return status;
+}
+
+std::vector<std::unique_ptr<cvc5::Command>> ExecutionContext::parseCommands(
+    parser::Parser* parser)
+{
+  std::vector<std::unique_ptr<cvc5::Command>> res;
+  while (true)
+  {
+    res.emplace_back(parser->nextCommand());
+    if (!res.back()) break;
+    if (dynamic_cast<cvc5::QuitCommand*>(res.back().get()) != nullptr)
+    {
+      break;
+    }
+  }
+  return res;
+}
+
+bool ExecutionContext::solveCommands(
+    std::vector<std::unique_ptr<cvc5::Command>>& cmds)
+{
+  bool interrupted = false;
+  bool status = true;
+  auto it = cmds.begin();
+  while (status && it != cmds.end())
+  {
+    if (interrupted)
+    {
+      solver().getDriverOptions().out() << cvc5::CommandInterrupted();
+      d_executor->reset();
+      break;
+    }
+
+    Command* cmd = it->get();
+
+    status = d_executor->doCommand(cmd);
+    if (cmd->interrupted() && status == 0)
+    {
+      interrupted = true;
+      break;
+    }
+
+    if (dynamic_cast<cvc5::QuitCommand*>(cmd) != nullptr)
+    {
+      break;
+    }
+  }
+  return status;
+}
+
+#if HAVE_SYS_WAIT_H
+
+namespace {
+
+/**
+ * Provides a convenient wrapper for POSIX pipes in the context of forking.
+ * The implemented mechanism is using a pipe to buffer the (standard or error)
+ * output of a child process and optionally copy it to the respective output of
+ * the parent process.
+ */
+class Pipe
+{
+ public:
+  /** Open a new pipe */
+  void open()
+  {
+    if (pipe(d_pipe) == -1)
+    {
+      throw internal::Exception("Unable to open pipe for child process");
+    }
+  }
+  /**
+   * Redirects the given file descriptor fd into this pipe using dup2 and closes
+   * both ends of the pipe. This method should be called within the child
+   * process after forking to redirect standard out or error out into the pipe.
+   */
+  void dup(int fd)
+  {
+    while ((dup2(d_pipe[1], fd) == -1) && (errno == EINTR))
+    {
+    }
```

Comment:
Added a comment. This was suggested in the tutorial I followed because `dup2` may be interrupted by a signal (`EINTR`) which is not really an error and just trying again should just work.